src/socket/mod.rs

```diff
@@ -43,18 +50,41 @@ impl Socket {
     /// This creates and binds a new UDP socket.
     // In general this function can be expanded to handle more advanced socket creation.
     pub(crate) async fn new_socket(
-        socket: &SocketAddr,
-    ) -> Result<tokio::net::UdpSocket, std::io::Error> {
-        tokio::net::UdpSocket::bind(socket).await
+        socket_addr: &SocketAddr,
+        ip_mode: IpMode,
+    ) -> Result<tokio::net::UdpSocket, Error> {
+        match ip_mode {
+            IpMode::Ip4 => match socket_addr {
+                SocketAddr::V6(_) => Err(Error::new(
+                    ErrorKind::InvalidInput,
+                    "Cannot create an ipv4 socket from an ipv6 address",
+                )),
+                ip4 => tokio::net::UdpSocket::bind(ip4).await,
+            },
+            IpMode::Ip6 {
+                enable_mapped_addresses,
+            } => {
+                let addr = match socket_addr {
+                    SocketAddr::V4(_) => Err(Error::new(
+                        ErrorKind::InvalidInput,
+                        "Cannot create an ipv6 socket from an ipv4 address",
+                    )),
+                    SocketAddr::V6(ip6) => Ok((*ip6).into()),
+                }?;
+                let socket = Socket2::new(Domain::IPV6, Type::DGRAM, Some(Protocol::UDP))?;
+                let only_v6 = !enable_mapped_addresses;
+                socket.set_only_v6(only_v6)?;
+                socket.set_nonblocking(true)?;
+                socket.bind(&addr)?;
+                tokio::net::UdpSocket::from_std(socket.into())
```

Comment:
I had issues doing this a while ago. 
Using a std socket caused some really funky issues and the solution was to not map from a std_socket. I've not looked into how to do this otherwise, or if its still an issue. 
See this: 
https://github.com/tokio-rs/tokio/issues/2969
and this:
https://github.com/tokio-rs/tokio/issues/3001

Reply:
The trick is the set non blocking. I haven't had any issues with that setup